Basic Station Integration: Spurious xtime jumps

#### Summary
For background on `xtime` and its handling in lorawan-stack read issue #1709.
Spurious 2**32-jumps of the `xtime` value sent to the gateway in the `dnmsg` message can be observed. This permanently puts gateway and server out of sync. As a consequence, downlinks scheduled based on the wrong xtime value are rejected by the Gateway because they are too far in the future. 

#### Steps to Reproduce

1. Start a gateway process and schedule a first downlink before the first 32 bit rollover.
2. Schedule another downlink some time after the first one, but still before the first 32 bit rollover.

#### What do you see now?
The first downlink has correct xtime. For the second downlink the `xtime` value is 2**32 larger than expected. Here is a concrete example:

```
<CONCENTRATOR START>
  "updf" xtime=00e1000000721200 (0)
...
  "updf" xtime=00e10000049e2b28 (0)
  "updf" xtime=00e100000536c120 (0)
 "dnmsg" xtime=00e100000536c120 (1) < -- first downlink OK
  "updf" xtime=00e1000005cf6b00 (0)
...
  "updf" xtime=00e100002b5c4050 (0)
  "updf" xtime=00e100002bf4d668 (0)
 "dnmsg" xtime=00e100012bf4d668 (1) < -- second downlink xtime offset by 2**32
  "updf" xtime=00e100002c8d6c40 (0)
 "dnmsg" xtime=00e100012c8d6c40 (1) < -- all following downlinks offset by 2**32
  "updf" xtime=00e100002d260220 (0)     /
 "dnmsg" xtime=00e100012d260220 (1) < -´
...
```

Here is the complete station.log:
[station.log](https://github.com/TheThingsNetwork/lorawan-stack/files/4667368/station.log)

The xtime log can be extracted from it like this
```bash
head -n 300 station.log  | sed -n 's/[^{]*\({.*}\).*/\1/p' | jq -cr '[.msgtype, (select(.msgtype == "updf") | .upinfo.xtime), (select(.msgtype == "dnmsg") | .xtime ) ,.RxDelay ] | @csv' | awk -F, '{ printf "%8.8s xtime=%016lx (%1d)\n", $1, $2, $3 }'
```

I suspect the issue is rooted in the tracking of 32 bit rollovers of `xtime`. In this particular case apparently a 32 bit rollover was assumed to have happened some time between `0x00000536c120` and `0x00002bf4d668` (~11 minutes).

#### What do you want to see instead?

The class A downlink `xtime` should not be offset by 2**32 to the corresponding uplink.

#### Environment
Basic Station 2.0.4
TTN community network

Found the cause. I can reproduce with a test case (in `pkg/gatewayserver/scheduling`):

```go
func TestIssue2581(t *testing.T) {
	a := assertions.New(t)

	clock := &RolloverClock{}

	timestamps := []int64{
		63331869818403100,
		63331869827372300,
		63331869837372200,
		63331869847372000,
		63331869857377000,
		63331869867371600,
		63331869877371600,
		63331869878424700,
		// snip
		63331870307366100,
		63331870317365900,
		63331870327365900,
		63331870337365800,
		63331870337361700, // Before the previous one
		63331870347365700,
		63331870357365600,
		63331870367365500,
	}
	var (
		prev      *int64
		sessionID int64
	)
	for i, xtimeIn := range timestamps {
		xtimeIn := xtimeIn
		diff := int64(0)
		if prev != nil {
			diff = xtimeIn - *prev
		}
		prev = &xtimeIn

		// timestamp can go back, but serverTime is always increasing
		timestamp := uint32(xtimeIn & 0xFFFFFFFF)
		serverTime := time.Now()

		if i == 0 {
			t.Log("Synchronizing gateway concentrator")
			sessionID = xtimeIn >> 48
			clock.SyncWithGatewayConcentrator(timestamp, serverTime, ConcentratorTime(time.Duration(xtimeIn&0xFFFFFFFFFFFF)*time.Microsecond))
		}
		rx := clock.Sync(timestamp, serverTime)
		tx := clock.FromTimestampTime(timestamp)

		t.Logf("xtimeIn=%016X tmst=%08X concentrator=%016X received=%v diff=%d", xtimeIn, timestamp, tx/1000, serverTime, diff)

		a.So(tx-rx, should.BeZeroValue)

		xtimeOut := sessionID<<48 | (int64(tx) / int64(time.Microsecond) & 0xFFFFFFFFFF)
		a.So(time.Duration(xtimeOut-xtimeIn)*time.Microsecond, should.BeZeroValue)
	}
}
```

Output:

```
=== RUN   TestIssue2581
    TestIssue2581: clock_test.go:166: Synchronizing gateway concentrator
    TestIssue2581: clock_test.go:173: xtimeIn=00E10000037CB91C tmst=037CB91C concentrator=00000000037CB91C received=2020-08-06 12:28:27.878791 +0200 CEST m=+0.009530010 diff=0
    TestIssue2581: clock_test.go:173: xtimeIn=00E100000405950C tmst=0405950C concentrator=000000000405950C received=2020-08-06 12:28:27.878962 +0200 CEST m=+0.009701407 diff=8969200
    TestIssue2581: clock_test.go:173: xtimeIn=00E10000049E2B28 tmst=049E2B28 concentrator=00000000049E2B28 received=2020-08-06 12:28:27.87897 +0200 CEST m=+0.009708853 diff=9999900
    TestIssue2581: clock_test.go:173: xtimeIn=00E100000536C0E0 tmst=0536C0E0 concentrator=000000000536C0E0 received=2020-08-06 12:28:27.878975 +0200 CEST m=+0.009713981 diff=9999800
    TestIssue2581: clock_test.go:173: xtimeIn=00E1000005CF6AE8 tmst=05CF6AE8 concentrator=0000000005CF6AE8 received=2020-08-06 12:28:27.878979 +0200 CEST m=+0.009718594 diff=10005000
    TestIssue2581: clock_test.go:173: xtimeIn=00E100000667EC50 tmst=0667EC50 concentrator=000000000667EC50 received=2020-08-06 12:28:27.878984 +0200 CEST m=+0.009723411 diff=9994600
    TestIssue2581: clock_test.go:173: xtimeIn=00E10000070082D0 tmst=070082D0 concentrator=00000000070082D0 received=2020-08-06 12:28:27.878999 +0200 CEST m=+0.009738467 diff=10000000
    TestIssue2581: clock_test.go:173: xtimeIn=00E100000710947C tmst=0710947C concentrator=000000000710947C received=2020-08-06 12:28:27.879005 +0200 CEST m=+0.009744058 diff=1053100
    TestIssue2581: clock_test.go:173: xtimeIn=00E1000020A1B4D4 tmst=20A1B4D4 concentrator=0000000020A1B4D4 received=2020-08-06 12:28:27.879009 +0200 CEST m=+0.009748619 diff=428941400
    TestIssue2581: clock_test.go:173: xtimeIn=00E10000213A4A8C tmst=213A4A8C concentrator=00000000213A4A8C received=2020-08-06 12:28:27.879014 +0200 CEST m=+0.009753177 diff=9999800
    TestIssue2581: clock_test.go:173: xtimeIn=00E1000021D2E10C tmst=21D2E10C concentrator=0000000021D2E10C received=2020-08-06 12:28:27.879018 +0200 CEST m=+0.009757585 diff=10000000
    TestIssue2581: clock_test.go:173: xtimeIn=00E10000226B7728 tmst=226B7728 concentrator=00000000226B7728 received=2020-08-06 12:28:27.879023 +0200 CEST m=+0.009761904 diff=9999900
    TestIssue2581: clock_test.go:173: xtimeIn=00E10000226B6724 tmst=226B6724 concentrator=00000001226B6724 received=2020-08-06 12:28:27.879042 +0200 CEST m=+0.009780997 diff=-4100
    TestIssue2581: doc.go:66: 
        /Users/johan/dev/src/lorawan-stack/pkg/gatewayserver/scheduling/clock_test.go:178
        '1h11m34.967296s' should have been the zero value
    TestIssue2581: clock_test.go:173: xtimeIn=00E1000023040D44 tmst=23040D44 concentrator=0000000123040D44 received=2020-08-06 12:28:27.879057 +0200 CEST m=+0.009796112 diff=10004000
    TestIssue2581: doc.go:66: 
        /Users/johan/dev/src/lorawan-stack/pkg/gatewayserver/scheduling/clock_test.go:178
        '1h11m34.967296s' should have been the zero value
    TestIssue2581: clock_test.go:173: xtimeIn=00E10000239CA360 tmst=239CA360 concentrator=00000001239CA360 received=2020-08-06 12:28:27.879067 +0200 CEST m=+0.009806553 diff=9999900
    TestIssue2581: doc.go:66: 
        /Users/johan/dev/src/lorawan-stack/pkg/gatewayserver/scheduling/clock_test.go:178
        '1h11m34.967296s' should have been the zero value
    TestIssue2581: clock_test.go:173: xtimeIn=00E100002435397C tmst=2435397C concentrator=000000012435397C received=2020-08-06 12:28:27.879106 +0200 CEST m=+0.009844876 diff=9999900
    TestIssue2581: doc.go:66: 
        /Users/johan/dev/src/lorawan-stack/pkg/gatewayserver/scheduling/clock_test.go:178
        '1h11m34.967296s' should have been the zero value
--- FAIL: TestIssue2581 (0.00s)
FAIL
exit status 1
FAIL    go.thethings.network/lorawan-stack/v3/pkg/gatewayserver/scheduling      0.287s
```

This is basically issue https://github.com/TheThingsNetwork/lorawan-stack/issues/1707. We should allow for small jumps back that are not considered roll overs.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -49,6 +49,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Troubleshooting guide.
 - API to get configuration from the Identity Server (including user registration options and password requirements).
 - Synchronize gateway time by uplink token on downstream in case the Gateway Server instance is not handling the upstream gateway connection.
+- Work-around for Basic Station gateways sending uplink frames with no `xtime`.
 
 ### Changed
 
@@ -84,6 +85,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Console MQTT URL validation.
 - AFCntDown from the application-layer is respected when skipping application payload crypto.
 - RTT usage for calculating downlink delta.
+- Synchronize concentrator timestamp when uplink messages arrive out-of-order.
 
 ## [3.8.6] - 2020-07-10
 

--- a/pkg/gatewayserver/io/basicstationlns/basicstationlns.go
+++ b/pkg/gatewayserver/io/basicstationlns/basicstationlns.go
@@ -431,6 +431,11 @@ func (s *srv) handleTraffic(c echo.Context) (err error) {
 				logger.WithError(err).Debug("Failed to parse join-request message")
 				return err
 			}
+			// TODO: Remove (https://github.com/lorabasics/basicstation/issues/74)
+			if jreq.UpInfo.XTime == 0 {
+				logger.Warn("Received join-request without xtime, drop message")
+				break
+			}
 			recordXTime(jreq.UpInfo.XTime, up.ReceivedAt)
 			if err := conn.HandleUp(up); err != nil {
 				logger.WithError(err).Warn("Failed to handle uplink message")
@@ -447,6 +452,11 @@ func (s *srv) handleTraffic(c echo.Context) (err error) {
 			if err != nil {
 				logger.WithError(err).Debug("Failed to parse uplink data frame")
 				return err
+			}
+			// TODO: Remove (https://github.com/lorabasics/basicstation/issues/74)
+			if updf.UpInfo.XTime == 0 {
+				logger.Warn("Received uplink data frame without xtime, drop message")
+				break
 			}
 			recordXTime(updf.UpInfo.XTime, up.ReceivedAt)
 			if err := conn.HandleUp(up); err != nil {

--- a/pkg/gatewayserver/io/basicstationlns/basicstationlns.go
+++ b/pkg/gatewayserver/io/basicstationlns/basicstationlns.go
@@ -334,20 +334,16 @@ func (s *srv) handleTraffic(c echo.Context) (err error) {
 		}
 	}()
 
-	var syncedConcentratorTime bool
 	recordXTime := func(timestamp int64, server time.Time) {
 		// The session is the 16 MSB.
 		atomic.StoreInt32(&sessionID, int32(timestamp>>48))
-		if !syncedConcentratorTime {
-			conn.SyncWithGatewayConcentrator(
-				// The concentrator timestamp is the 32 LSB.
-				uint32(timestamp&0xFFFFFFFF),
-				server,
-				// The Basic Station epoch is the 48 LSB.
-				scheduling.ConcentratorTime(time.Duration(timestamp&0xFFFFFFFFFF)*time.Microsecond),
-			)
-			syncedConcentratorTime = true
-		}
+		conn.SyncWithGatewayConcentrator(
+			// The concentrator timestamp is the 32 LSB.
+			uint32(timestamp&0xFFFFFFFF),
+			server,
+			// The Basic Station epoch is the 48 LSB.
+			scheduling.ConcentratorTime(time.Duration(timestamp&0xFFFFFFFFFF)*time.Microsecond),
+		)
 	}
 
 	for {

--- a/pkg/gatewayserver/scheduling/clock.go
+++ b/pkg/gatewayserver/scheduling/clock.go
@@ -59,10 +59,13 @@ func (c *RolloverClock) SyncTime() (time.Time, bool) {
 // Sync synchronizes the clock with the given concentrator timestamp and the server time.
 func (c *RolloverClock) Sync(timestamp uint32, server time.Time) ConcentratorTime {
 	rollovers := int64(c.absolute/ConcentratorTime(time.Microsecond)) >> 32
-	if passed := int64(timestamp) - int64(c.relative); passed < 0 {
+	// If the new timestamp is lower, it is considered a roll over if it advanced more than 25%.
+	// Otherwise, this is considered an out-of-order arrival.
+	if passed := int64(timestamp) - int64(c.relative); passed < -0x40000000 {
 		rollovers++
 	}
-	if c.server != nil {
+	// If there are full roll over windows between the last and the current server time, take these into account.
+	if c.server != nil && server.After(*c.server) {
 		rollovers += int64(server.Sub(*c.server)/time.Microsecond) >> 32
 	}
 	c.absolute = (ConcentratorTime(rollovers<<32) + ConcentratorTime(timestamp)) * ConcentratorTime(time.Microsecond)

--- a/pkg/gatewayserver/scheduling/clock_test.go
+++ b/pkg/gatewayserver/scheduling/clock_test.go
@@ -33,8 +33,8 @@ func TestRolloverClock(t *testing.T) {
 		Relative uint32
 	}{
 		{
-			Absolute: ConcentratorTime(10 * time.Second),
-			Relative: uint32(10000000),
+			Absolute: ConcentratorTime(20 * time.Minute),
+			Relative: uint32(20 * time.Minute / time.Microsecond),
 		},
 		{
 			// 1 rollover.
@@ -43,8 +43,8 @@ func TestRolloverClock(t *testing.T) {
 		},
 		{
 			// 3 rollovers (1 existing + 2 server time rollovers).
-			Absolute: ConcentratorTime(3<<32*time.Microsecond) + ConcentratorTime(10*time.Second),
-			Relative: uint32(10000000),
+			Absolute: ConcentratorTime(3<<32*time.Microsecond) + ConcentratorTime(30*time.Minute),
+			Relative: uint32(30 * time.Minute / time.Microsecond),
 		},
 		{
 			// 5 rollovers (3 existing + 1 concentrator timestamp rollover + 1 server time rollover).
@@ -120,4 +120,62 @@ func TestSyncWithGatewayConcentrator(t *testing.T) {
 	clock.SyncWithGatewayConcentrator(0x496054D6, time.Now(), ConcentratorTime(0xAA496054D6)*ConcentratorTime(time.Microsecond))
 	v := int64(clock.FromTimestampTime(0x499D5DD6)) / int64(time.Microsecond)
 	a.So(v, should.Equal, int64(0xAA499D5DD6))
+}
+
+// TestIssue2581 is a test case for resolving https://github.com/TheThingsNetwork/lorawan-stack/issues/2581.
+func TestIssue2581(t *testing.T) {
+	a := assertions.New(t)
+
+	clock := &RolloverClock{}
+
+	timestamps := []int64{
+		63331869818403100,
+		63331869827372300,
+		63331869837372200,
+		63331869847372000,
+		63331869857377000,
+		63331869867371600,
+		63331869877371600,
+		63331869878424700,
+		// snip
+		63331870307366100,
+		63331870317365900,
+		63331870327365900,
+		63331870337365800,
+		63331870337361700, // Before the previous one
+		63331870347365700,
+		63331870357365600,
+		63331870367365500,
+	}
+	var (
+		prev      *int64
+		sessionID int64
+	)
+	for i, xtimeIn := range timestamps {
+		xtimeIn := xtimeIn
+		diff := int64(0)
+		if prev != nil {
+			diff = xtimeIn - *prev
+		}
+		prev = &xtimeIn
+
+		// timestamp can go back, but serverTime is always increasing
+		timestamp := uint32(xtimeIn & 0xFFFFFFFF)
+		serverTime := time.Now()
+
+		if i == 0 {
+			t.Log("Synchronizing gateway concentrator")
+			sessionID = xtimeIn >> 48
+			clock.SyncWithGatewayConcentrator(timestamp, serverTime, ConcentratorTime(time.Duration(xtimeIn&0xFFFFFFFFFFFF)*time.Microsecond))
+		}
+		rx := clock.Sync(timestamp, serverTime)
+		tx := clock.FromTimestampTime(timestamp)
+
+		t.Logf("xtimeIn=%016X tmst=%08X concentrator=%016X received=%v diff=%d", xtimeIn, timestamp, tx/1000, serverTime, diff)
+
+		a.So(tx-rx, should.BeZeroValue)
+
+		xtimeOut := sessionID<<48 | (int64(tx) / int64(time.Microsecond) & 0xFFFFFFFFFF)
+		a.So(time.Duration(xtimeOut-xtimeIn)*time.Microsecond, should.BeZeroValue)
+	}
 }

